File: backend/app/channels/telegram.py
from typing import Any

import httpx

from app.channels.base import InboundMessage


class TelegramAdapter:
    channel = "TELEGRAM"

    def __init__(self, bot_token: str | None) -> None:
        self._bot_token = bot_token

    def normalize(self, payload: dict[str, Any]) -> InboundMessage | None:
        update_id = payload.get("update_id")
        message = payload.get("message") or payload.get("edited_message")
        if update_id is None or not isinstance(message, dict):
            return None
        sender = message.get("from")
        chat = message.get("chat")
        if not isinstance(sender, dict) or not isinstance(chat, dict):
            return None
        sender_id = sender.get("id")
        chat_id = chat.get("id")
        message_id = message.get("message_id")
        if sender_id is None or chat_id is None or message_id is None:
            return None
        first_name = str(sender.get("first_name") or "").strip()
        last_name = str(sender.get("last_name") or "").strip()
        display_name = " ".join(part for part in (first_name, last_name) if part) or None
        attachments: list[dict[str, Any]] = []
        if message.get("document"):
            document = message["document"]
            attachments.append(
                {
                    "kind": "document",
                    "provider_file_id": document.get("file_id"),
                    "filename": document.get("file_name"),
                    "mime_type": document.get("mime_type"),
                }
            )
        if message.get("photo"):
            photo = message["photo"][-1]
            attachments.append(
                {
                    "kind": "photo",
                    "provider_file_id": photo.get("file_id"),
                }
            )
        return InboundMessage(
            provider_event_id=str(update_id),
            provider_message_id=str(message_id),
            external_user_id=str(sender_id),
            external_chat_id=str(chat_id),
            username=str(sender.get("username")) if sender.get("username") else None,
            display_name=display_name,
            text=str(message.get("text") or message.get("caption") or "").strip(),
            attachments=attachments,
        )

    async def send_text(
        self, external_chat_id: str, text: str, *, idempotency_key: str
    ) -> str:
        if not self._bot_token:
            raise RuntimeError("TELEGRAM_BOT_TOKEN is not configured")
        async with httpx.AsyncClient(timeout=20.0) as client:
            response = await client.post(
                f"https://api.telegram.org/bot{self._bot_token}/sendMessage",
                json={
                    "chat_id": external_chat_id,
                    "text": text,
                    "disable_web_page_preview": True,
                },
                headers={"X-IMMIGROME-Idempotency-Key": idempotency_key},
            )
        response.raise_for_status()
        payload: dict[str, Any] = response.json()
        if not payload.get("ok") or not isinstance(payload.get("result"), dict):
            raise RuntimeError("Telegram returned an invalid sendMessage response")
        return str(payload["result"]["message_id"])

File: src/server/channels/telegram.ts
import { createHash } from "crypto";

export type TelegramAttachment = {
  kind: "document" | "photo";
  providerFileId?: string;
  filename?: string;
  mimeType?: string;
};

export type NormalizedTelegramMessage = {
  kind: "message";
  providerEventId: string;
  providerMessageId: string;
  externalUserId: string;
  externalChatId: string;
  username: string | null;
  displayName: string | null;
  text: string;
  attachments: TelegramAttachment[];
};

export type NormalizedTelegramCallback = {
  kind: "callback";
  providerEventId: string;
  callbackQueryId: string;
  providerMessageId: string | null;
  externalUserId: string;
  externalChatId: string;
  username: string | null;
  displayName: string | null;
  data: string;
};

export type NormalizedTelegramUpdate =
  | NormalizedTelegramMessage
  | NormalizedTelegramCallback;

/**
 * Normalize a Telegram Bot API Update into a channel-agnostic inbound message
 * or appointment callback. Returns null for updates we ignore.
 */
export function normalizeTelegramUpdate(
  payload: unknown,
): NormalizedTelegramUpdate | null {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    return null;
  }
  const update = payload as Record<string, unknown>;
  const updateId = update.update_id;
  if (updateId == null) return null;

  const callback = update.callback_query as Record<string, unknown> | undefined;
  if (callback && typeof callback === "object") {
    return normalizeCallback(updateId, callback);
  }

  const message =
    (update.message as Record<string, unknown> | undefined) ??
    (update.edited_message as Record<string, unknown> | undefined);
  if (!message || typeof message !== "object") return null;

  const sender = message.from;
  const chat = message.chat;
  if (!sender || typeof sender !== "object" || Array.isArray(sender)) return null;
  if (!chat || typeof chat !== "object" || Array.isArray(chat)) return null;

  const senderObj = sender as Record<string, unknown>;
  const chatObj = chat as Record<string, unknown>;
  const senderId = senderObj.id;
  const chatId = chatObj.id;
  const messageId = message.message_id;
  if (senderId == null || chatId == null || messageId == null) return null;

  const firstName = String(senderObj.first_name ?? "").trim();
  const lastName = String(senderObj.last_name ?? "").trim();
  const displayName =
    [firstName, lastName].filter(Boolean).join(" ").trim() || null;

  const attachments: TelegramAttachment[] = [];
  const document = message.document;
  if (document && typeof document === "object" && !Array.isArray(document)) {
    const doc = document as Record<string, unknown>;
    attachments.push({
      kind: "document",
      providerFileId: doc.file_id != null ? String(doc.file_id) : undefined,
      filename: doc.file_name != null ? String(doc.file_name) : undefined,
      mimeType: doc.mime_type != null ? String(doc.mime_type) : undefined,
    });
  }
  const photo = message.photo;
  if (Array.isArray(photo) && photo.length > 0) {
    const last = photo[photo.length - 1] as Record<string, unknown>;
    attachments.push({
      kind: "photo",
      providerFileId: last.file_id != null ? String(last.file_id) : undefined,
    });
  }

  return {
    kind: "message",
    providerEventId: String(updateId),
    providerMessageId: String(messageId),
    externalUserId: String(senderId),
    externalChatId: String(chatId),
    username: senderObj.username != null ? String(senderObj.username) : null,
    displayName,
    text: String(message.text ?? message.caption ?? "").trim(),
    attachments,
  };
}

function normalizeCallback(
  updateId: unknown,
  callback: Record<string, unknown>,
): NormalizedTelegramCallback | null {
  const callbackId = callback.id;
  const data = String(callback.data ?? "").trim();
  if (callbackId == null || !data) return null;

  const sender = callback.from;
  if (!sender || typeof sender !== "object" || Array.isArray(sender)) return null;
  const senderObj = sender as Record<string, unknown>;
  const senderId = senderObj.id;
  if (senderId == null) return null;

  const message = callback.message as Record<string, unknown> | undefined;
  const chat = message?.chat as Record<string, unknown> | undefined;
  const chatId = chat?.id ?? null;
  if (chatId == null) return null;

  const firstName = String(senderObj.first_name ?? "").trim();
  const lastName = String(senderObj.last_name ?? "").trim();
  const displayName =
    [firstName, lastName].filter(Boolean).join(" ").trim() || null;

  return {
    kind: "callback",
    providerEventId: String(updateId),
    callbackQueryId: String(callbackId),
    providerMessageId:
      message?.message_id != null ? String(message.message_id) : null,
    externalUserId: String(senderId),
    externalChatId: String(chatId),
    username: senderObj.username != null ? String(senderObj.username) : null,
    displayName,
    data,
  };
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(",")}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  return `{${keys
    .map((key) => `${JSON.stringify(key)}:${stableStringify(obj[key])}`)
    .join(",")}}`;
}

/** Canonical hash with sorted keys for InboxEvent.payloadHash. */
export function hashJson(value: unknown): string {
  return createHash("sha256").update(stableStringify(value)).digest("hex");
}

/** Parse appointment callback_data: a:{id}:{token}:{action}[:slotKey] */
export function parseAppointmentCallbackData(data: string): {
  appointmentId: string;
  token: string;
  action: "ok" | "alt" | "x" | "s";
  slotKey?: string;
} | null {
  const parts = data.split(":");
  if (parts[0] !== "a" || parts.length < 4) return null;
  const appointmentId = parts[1];
  const token = parts[2];
  const action = parts[3];
  if (!appointmentId || !token) return null;
  if (action === "ok" || action === "alt" || action === "x") {
    return { appointmentId, token, action };
  }
  if (action === "s" && parts[4]) {
    return { appointmentId, token, action: "s", slotKey: parts.slice(4).join(":") };
  }
  return null;
}
